File: app/src/main/java/com/yu/functionbox/main/MainViewModel.java
package com.yu.functionbox.main;

import android.databinding.ObservableArrayList;
import android.databinding.ObservableList;
import android.support.annotation.NonNull;
import android.text.TextUtils;
import android.util.Log;

import com.yu.functionbox.base.BaseBizResult;
import com.yu.functionbox.base.BizCallback;
import com.yu.functionbox.base.BizResult;
import com.yu.functionbox.data.FunctionBean;
import com.yu.functionbox.data.SceneBean;
import com.yu.functionbox.data.SortBean;
import com.yu.functionbox.db.DbService;
import com.yu.functionbox.event.EventMessage;
import com.yu.functionbox.event.MyEvent;
import com.yu.functionbox.utils.EventBusUtils;
import com.yu.functionbox.utils.ToastUtils;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;

import java.util.ArrayList;

/**
 * Created by yuw on 2017-12-29.
 * description
 */

public class MainViewModel {
    private final static String TAG = "MainViewModel";
    public ObservableList<SortBean> mSortBeans = new ObservableArrayList<>();
    public ObservableList<SceneBean> mSceneBeans = new ObservableArrayList<>();
    public ObservableList<FunctionBean> mFunctionBeans = new ObservableArrayList<>();
    private long mSortId;
    private long mSceneId;

    public MainViewModel() {
        EventBusUtils.register(this);
    }

    public void initData() {
        DbService.get().getSorts(new BizCallback<ArrayList<SortBean>>() {
            @Override
            public void onSucceed(ArrayList<SortBean> response, @NonNull BizResult bizResult) {
                mSortBeans.clear();
                mSortBeans.addAll(response);
                if (response.size() > 0) {
                    EventBus.getDefault().post(new EventMessage<>(MyEvent.EVENT_SELECT_FIRST_SORT));
                    long id = response.get(0).getId();
                    mSortId = id;
                    getScenesBySortId(id);
                }
            }

            @Override
            public void onFailed(String errorMsg, @NonNull BizResult bizResult) {
                Log.i(TAG, "getSorts onFailed: ");
            }

        });

    }

    private void getScenesBySortId(long id) {
        DbService.get().getScenes(id, new BizCallback<ArrayList<SceneBean>>() {
            @Override
            public void onSucceed(ArrayList<SceneBean> response, @NonNull BizResult bizResult) {
                mSceneBeans.clear();
                mSceneBeans.addAll(response);
                if (response.size() > 0) {
                    EventBus.getDefault().post(new EventMessage<>(MyEvent.EVENT_SELECT_FIRST_SCENE));
                    mSceneId = response.get(0).getId();
                    getFunctionBySceneId(mSceneId);
                } else {
                    mFunctionBeans.clear();
                }

            }

            @Override
            public void onFailed(String errorMsg, @NonNull BizResult bizResult) {
                Log.i(TAG, "getScenesBySortId onFailed: ");
            }

        });
    }

    private void getFunctionBySceneId(long sceneId) {
        DbService.get().getFunctions(sceneId, new BizCallback<ArrayList<FunctionBean>>() {
            @Override
            public void onSucceed(ArrayList<FunctionBean> response, @NonNull BizResult bizResult) {
                mFunctionBeans.clear();
                mFunctionBeans.addAll(response);
            }

            @Override
            public void onFailed(String errorMsg, @NonNull BizResult bizResult) {
                Log.i(TAG, "getFunctionBySceneId onFailed: ");
            }

        });
    }


    @Subscribe
    public void onEventUpdate(EventMessage event) {
        switch (event.mCode) {
            case MyEvent.EVENT_INSERT_SORT:
                String title = (String) event.mObj;
                if (!TextUtils.isEmpty(title)) {
                    DbService.get().insertSort(title, new BizCallback() {
                        @Override
                        public void onSucceed(Object response, @NonNull BaseBizResult bizResult) {
                            ToastUtils.show("添加成功!");
                            updateSort();
                        }

                        @Override
                        public void onFailed(String errorMsg, @NonNull BaseBizResult bizResult) {

                        }
                    });
                }
                break;
            case MyEvent.EVENT_INSERT_SCENE:
                String title2 = (String) event.mObj;
                if (!TextUtils.isEmpty(title2)) {
                    DbService.get().insertScenes(mSortId, title2, new BizCallback() {
                        @Override
                        public void onSucceed(Object response, @NonNull BaseBizResult bizResult) {
                            ToastUtils.show("添加成功!");
                            updateScene();
                        }

                        @Override
                        public void onFailed(String errorMsg, @NonNull BaseBizResult bizResult) {

                        }
                    });
                }
                break;
            case MyEvent.EVENT_INSERT_FUNCTION:
                String detail = (String) event.mObj;
                if (!TextUtils.isEmpty(detail)) {
                    DbService.get().insertFunction(mSceneId, "test", detail, new BizCallback() {
                        @Override
                        public void onSucceed(Object response, @NonNull BaseBizResult bizResult) {
                            ToastUtils.show("添加成功!");
                            updateFunction();
                        }

                        @Override
                        public void onFailed(String errorMsg, @NonNull BaseBizResult bizResult) {

                        }
                    });
                }
                break;
            case MyEvent.EVENT_CLICK_SORT:
                mSortId = (long) event.mObj;
                getScenesBySortId(mSortId);
                break;
            case MyEvent.EVENT_CLICK_SCENE:
                mSceneId = (long) event.mObj;
                getFunctionBySceneId(mSceneId);
                break;
            case MyEvent.EVENT_SAVE_FUNCTION:
                saveFunction((FunctionBean) event.mObj);
                break;
        }
    }

    private void saveFunction(FunctionBean bean) {
        DbService.get().updateFunction(bean.getId(), bean.getDetail(), new BizCallback() {
            @Override
            public void onSucceed(Object response, @NonNull BaseBizResult bizResult) {
                ToastUtils.show("保存成功");
            }

            @Override
            public void onFailed(String errorMsg, @NonNull BaseBizResult bizResult) {
                ToastUtils.show("保存失败");
            }
        });
    }


    private void updateSort() {
        DbService.get().getSorts(new BizCallback<ArrayList<SortBean>>() {
            @Override
            public void onSucceed(ArrayList<SortBean> response, @NonNull BizResult bizResult) {
                mSortBeans.clear();
                mSortBeans.addAll(response);
                if (response.size() > 0) {
                    EventBus.getDefault().post(new EventMessage<>(MyEvent.EVENT_SELECT_LAST_SORT, response.size()));
                    long id = response.get(response.size()).getId();
                    mSortId = id;
                    getScenesBySortId(id);
                }
            }

            @Override
            public void onFailed(String errorMsg, @NonNull BizResult bizResult) {
                Log.i(TAG, "getSorts onFailed: ");
            }

        });
    }

    private void updateScene() {
        DbService.get().getScenes(mSortId, new BizCallback<ArrayList<SceneBean>>() {
            @Override
            public void onSucceed(ArrayList<SceneBean> response, @NonNull BizResult bizResult) {
                mSceneBeans.clear();
                mSceneBeans.addAll(response);
                EventBus.getDefault().post(new EventMessage<>(MyEvent.EVENT_SELECT_LAST_SCENE, response.size()));
                if (response.size() > 0) {
                    mSceneId = response.get(response.size() - 1).getId();
                    getFunctionBySceneId(mSceneId);
                }
            }

            @Override
            public void onFailed(String errorMsg, @NonNull BizResult bizResult) {
                Log.i(TAG, "getScenesBySortId onFailed: ");
            }

        });
    }

    public void updateFunction() {
        DbService.get().getFunctions(mSceneId, new BizCallback<ArrayList<FunctionBean>>() {
            @Override
            public void onSucceed(ArrayList<FunctionBean> response, @NonNull BizResult bizResult) {
                mFunctionBeans.clear();
                mFunctionBeans.addAll(response);
            }

            @Override
            public void onFailed(String errorMsg, @NonNull BizResult bizResult) {
                Log.i(TAG, "getFunctionBySceneId onFailed: ");
            }

        });
    }

    public void destroy() {
        EventBusUtils.unRegister(this);
    }
}

File: app/src/main/java/com/yu/functionbox/main/MainViewModel.kt
package com.yu.functionbox.main

import android.databinding.ObservableArrayList
import android.databinding.ObservableList
import android.text.TextUtils
import android.util.Log
import com.yu.functionbox.base.BizCallback
import com.yu.functionbox.base.BizResult
import com.yu.functionbox.data.FunctionBean
import com.yu.functionbox.data.SceneBean
import com.yu.functionbox.data.SortBean
import com.yu.functionbox.db.DbService
import com.yu.functionbox.event.EventMessage
import com.yu.functionbox.event.MyEvent
import com.yu.functionbox.utils.EventBusUtils
import com.yu.functionbox.utils.ToastUtils
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import java.util.*

/**
 * Created by yuw on 2017-12-29.
 * description
 */

class MainViewModel {
    var mSortBeans: ObservableList<SortBean> = ObservableArrayList()
    var mSceneBeans: ObservableList<SceneBean> = ObservableArrayList()
    var mFunctionBeans: ObservableList<FunctionBean> = ObservableArrayList()
    private var mSortId: Long = 0
    private var mSceneId: Long = 0

    init {
        EventBusUtils.register(this)
    }

    fun initData() {
        DbService.getSorts(object : BizCallback<ArrayList<SortBean>> {
            override fun onSucceed(response: ArrayList<SortBean>, bizResult: BizResult) {
                mSortBeans.clear()
                mSortBeans.addAll(response)
                if (response.size > 0) {
                    EventBus.getDefault().post(EventMessage<Any>(MyEvent.EVENT_SELECT_FIRST_SORT))
                    val id = response[0].id
                    mSortId = id
                    getScenesBySortId(id)
                }
            }

            override fun onFailed(errorMsg: String, bizResult: BizResult) {
                Log.i(TAG, "getSorts onFailed: ")
            }

        })

    }

    private fun getScenesBySortId(id: Long) {
        DbService.getScenes(id, object : BizCallback<ArrayList<SceneBean>> {
            override fun onSucceed(response: ArrayList<SceneBean>, bizResult: BizResult) {
                mSceneBeans.clear()
                mSceneBeans.addAll(response)
                if (response.size > 0) {
                    EventBus.getDefault().post(EventMessage<Any>(MyEvent.EVENT_SELECT_FIRST_SCENE))
                    mSceneId = response[0].id
                    getFunctionBySceneId(mSceneId)
                } else {
                    mFunctionBeans.clear()
                }

            }

            override fun onFailed(errorMsg: String, bizResult: BizResult) {
                Log.i(TAG, "getScenesBySortId onFailed: ")
            }

        })
    }

    private fun getFunctionBySceneId(sceneId: Long) {
        DbService.getFunctions(sceneId, object : BizCallback<ArrayList<FunctionBean>> {
            override fun onSucceed(response: ArrayList<FunctionBean>, bizResult: BizResult) {
                mFunctionBeans.clear()
                mFunctionBeans.addAll(response)
            }

            override fun onFailed(errorMsg: String, bizResult: BizResult) {
                Log.i(TAG, "getFunctionBySceneId onFailed: ")
            }

        })
    }


    @Subscribe
    fun onEventUpdate(event: EventMessage<*>) {
        when (event.mCode) {
            MyEvent.EVENT_INSERT_SORT -> {
                val title = event.mObj as String
                if (!TextUtils.isEmpty(title)) {
                    DbService.insertSort(title, object : BizCallback<Any> {
                        override fun onSucceed(response: Any?, bizResult: BizResult) {
                            ToastUtils.show("添加成功!")
                            updateSort()
                        }

                        override fun onFailed(errorMsg: String?, bizResult: BizResult) {

                        }
                    })
                }
            }
            MyEvent.EVENT_INSERT_SCENE -> {
                val title2 = event.mObj as String
                if (!TextUtils.isEmpty(title2)) {
                    DbService.insertScenes(mSortId, title2, object : BizCallback<Any> {
                        override fun onSucceed(response: Any?, bizResult: BizResult) {
                            ToastUtils.show("添加成功!")
                            updateScene()
                        }

                        override fun onFailed(errorMsg: String?, bizResult: BizResult) {

                        }
                    })
                }
            }
            MyEvent.EVENT_INSERT_FUNCTION -> {
                val detail = event.mObj as String
                if (!TextUtils.isEmpty(detail)) {
                    DbService.insertFunction(mSceneId, "test", detail, object : BizCallback<Any> {
                        override fun onSucceed(response: Any?, bizResult: BizResult) {
                            ToastUtils.show("添加成功!")
                            updateFunction()
                        }

                        override fun onFailed(errorMsg: String, bizResult: BizResult) {

                        }
                    })
                }
            }
            MyEvent.EVENT_CLICK_SORT -> {
                mSortId = event.mObj as Long
                getScenesBySortId(mSortId)
            }
            MyEvent.EVENT_CLICK_SCENE -> {
                mSceneId = event.mObj as Long
                getFunctionBySceneId(mSceneId)
            }
            MyEvent.EVENT_SAVE_FUNCTION -> saveFunction(event.mObj as FunctionBean)
        }
    }

    private fun saveFunction(bean: FunctionBean) {
        DbService.updateFunction(bean.id, bean.detail, object : BizCallback<Any> {
            override fun onSucceed(response: Any?, bizResult: BizResult) {
                ToastUtils.show("保存成功")
            }

            override fun onFailed(errorMsg: String, bizResult: BizResult) {
                ToastUtils.show("保存失败")
            }
        })
    }


    private fun updateSort() {
        DbService.getSorts(object : BizCallback<ArrayList<SortBean>> {
            override fun onSucceed(response: ArrayList<SortBean>, bizResult: BizResult) {
                mSortBeans.clear()
                mSortBeans.addAll(response)
                if (response.size > 0) {
                    EventBus.getDefault().post(EventMessage(MyEvent.EVENT_SELECT_LAST_SORT, response.size))
                    val id = response[response.size].id
                    mSortId = id
                    getScenesBySortId(id)
                }
            }

            override fun onFailed(errorMsg: String, bizResult: BizResult) {
                Log.i(TAG, "getSorts onFailed: ")
            }

        })
    }

    private fun updateScene() {
        DbService.getScenes(mSortId, object : BizCallback<ArrayList<SceneBean>> {
            override fun onSucceed(response: ArrayList<SceneBean>, bizResult: BizResult) {
                mSceneBeans.clear()
                mSceneBeans.addAll(response)
                EventBus.getDefault().post(EventMessage(MyEvent.EVENT_SELECT_LAST_SCENE, response.size))
                if (response.size > 0) {
                    mSceneId = response[response.size - 1].id
                    getFunctionBySceneId(mSceneId)
                }
            }

            override fun onFailed(errorMsg: String, bizResult: BizResult) {
                Log.i(TAG, "getScenesBySortId onFailed: ")
            }

        })
    }

    fun updateFunction() {
        DbService.getFunctions(mSceneId, object : BizCallback<ArrayList<FunctionBean>> {
            override fun onSucceed(response: ArrayList<FunctionBean>, bizResult: BizResult) {
                mFunctionBeans.clear()
                mFunctionBeans.addAll(response)
            }

            override fun onFailed(errorMsg: String, bizResult: BizResult) {
                Log.i(TAG, "getFunctionBySceneId onFailed: ")
            }

        })
    }

    fun destroy() {
        EventBusUtils.unRegister(this)
    }

    companion object {
        private val TAG = "MainViewModel"
    }
}
